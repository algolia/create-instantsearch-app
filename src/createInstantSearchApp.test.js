const createInstantSearchAppFactory = require('./createInstantSearchApp');

let buildAppSpy;
let fetchLibraryVersionsSpy;
let installDependenciesSpy;

const createInstantSearchApp = (path, config) => {
  buildAppSpy = jest.fn(() => Promise.resolve());
  fetchLibraryVersionsSpy = jest.fn();
  installDependenciesSpy = jest.fn();

  return createInstantSearchAppFactory(path, config, {
    buildApp: buildAppSpy,
    fetchLibraryVersions: fetchLibraryVersionsSpy,
    installDependencies: installDependenciesSpy,
  });
};

describe('Config', () => {
  describe('required', () => {
    test('without path throws', () => {
      const createApp = () => createInstantSearchApp('', {});

      expect(createApp).toThrowErrorMatchingSnapshot();
    });

    test('without template throws', () => {
      const createApp = () => createInstantSearchApp('/tmp/test-app', {});

      expect(createApp).toThrowErrorMatchingSnapshot();
    });

    test('with unknown template throws', () => {
      const createApp = () =>
        createInstantSearchApp('/tmp/test-app', {
          template: 'UnknownTemplate',
        });

      expect(createApp).toThrowErrorMatchingSnapshot();
    });

    test('with correct template does not throw', () => {
      const createApp = () =>
        createInstantSearchApp('/tmp/test-app', {
          template: 'InstantSearch.js',
        });

      expect(createApp).not.toThrow();
    });

    test('with path and template does not throw', () => {
      const createApp = () =>
        createInstantSearchApp('/tmp/test-app', {
          template: 'InstantSearch.js',
        });

      expect(createApp).not.toThrow();
    });
  });

  describe('buildApp', () => {
    test('gets called', () => {
      createInstantSearchApp('/tmp/test-app', {
        template: 'InstantSearch.js',
      });

      expect(buildAppSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('installDependencies', () => {
    test('with installation undefined installs the dependencies gets called', async () => {
      expect.assertions(1);

      await createInstantSearchApp('/tmp/test-app', {
        template: 'InstantSearch.js',
      });

      expect(installDependenciesSpy).toHaveBeenCalledTimes(1);
    });

    test('with installation installs the dependencies gets called', async () => {
      expect.assertions(1);

      await createInstantSearchApp('/tmp/test-app', {
        template: 'InstantSearch.js',
        installation: true,
      });

      expect(installDependenciesSpy).toHaveBeenCalledTimes(1);
    });

    test('without installation does not install the dependencies gets called', async () => {
      expect.assertions(1);

      await createInstantSearchApp('/tmp/test-app', {
        template: 'InstantSearch.js',
        installation: false,
      });

      expect(installDependenciesSpy).not.toHaveBeenCalled();
    });
  });
});
