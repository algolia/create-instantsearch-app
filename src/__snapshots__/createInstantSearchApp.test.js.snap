// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config required with unknown template throws 1`] = `"The option \`template\` is required."`;

exports[`Config required without path throws 1`] = `"The option \`path\` is required."`;

exports[`Config required without template throws 1`] = `"The option \`template\` is required."`;
